File: example/cpp/grpc_server.cpp
#include <iostream>
#include <memory>
#include <string>
#include <grpcpp/grpcpp.h>
#include "../../generated/cpp/uav_service.pb.h"
#include "../../generated/cpp/uav_service.grpc.pb.h"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;
using uav::UavService;
using uav::UploadStatusRequest;
using uav::UploadStatusResponse;
using uav::SetSafetySpaceRequest;
using uav::SetSafetySpaceResponse;

// ============================================================================
// UavService 服务实现
// ============================================================================

class UavServiceImpl final : public UavService::Service {
public:
    /**
     * 处理上传无人机状态数据的请求
     */
    Status UploadStatus(ServerContext* /*context*/,
                       const UploadStatusRequest* request,
                       UploadStatusResponse* response) override {
        std::cout << "[服务端] 收到上传状态请求 - UAV ID: " << request->uav_id()
                  << ", 数据长度: " << request->data().size() << " 字节" << std::endl;
        
        // 数据验证
        if (request->uav_id().empty()) {
            response->set_success(false);
            response->set_message("uav_id 不能为空");
            return Status(grpc::StatusCode::INVALID_ARGUMENT, "uav_id 不能为空");
        }
        
        if (request->data().empty()) {
            response->set_success(false);
            response->set_message("data 不能为空");
            return Status(grpc::StatusCode::INVALID_ARGUMENT, "data 不能为空");
        }
        
        // TODO: 在这里添加你的业务逻辑
        // 例如：
        // - 解析和验证数据
        // - 保存到数据库
        // - 更新无人机状态
        // - 触发其他服务
        
        std::cout << "[服务端] 成功处理 UAV " << request->uav_id() << " 的状态数据" << std::endl;
        
        response->set_success(true);
        response->set_message("成功接收 UAV " + request->uav_id() + 
                             " 的状态数据，数据大小: " + 
                             std::to_string(request->data().size()) + " 字节");
        
        return Status::OK;
    }
    
    /**
     * 处理下发无人机安全空间的请求
     */
    Status SetSafetySpace(ServerContext* /*context*/,
                         const SetSafetySpaceRequest* request,
                         SetSafetySpaceResponse* response) override {
        std::cout << "[服务端] 收到设置安全空间请求 - UAV ID: " << request->uav_id()
                  << ", 数据长度: " << request->data().size() << " 字节" << std::endl;
        
        // 数据验证
        if (request->uav_id().empty()) {
            response->set_success(false);
            response->set_message("uav_id 不能为空");
            return Status(grpc::StatusCode::INVALID_ARGUMENT, "uav_id 不能为空");
        }
        
        if (request->data().empty()) {
            response->set_success(false);
            response->set_message("data 不能为空");
            return Status(grpc::StatusCode::INVALID_ARGUMENT, "data 不能为空");
        }
        
        // TODO: 在这里添加你的业务逻辑
        // 例如：
        // - 解析和验证安全空间数据
        // - 检查权限
        // - 下发到无人机
        // - 更新配置数据库
        // - 通知相关服务
        
        std::cout << "[服务端] 成功为 UAV " << request->uav_id() << " 设置安全空间" << std::endl;
        
        response->set_success(true);
        response->set_message("成功为 UAV " + request->uav_id() + 
                             " 设置安全空间，数据大小: " + 
                             std::to_string(request->data().size()) + " 字节");
        
        return Status::OK;
    }
};

// ============================================================================
// 服务器启动函数
// ============================================================================

void RunServer(int port = 50051) {
    std::string server_address("0.0.0.0:" + std::to_string(port));
    UavServiceImpl service;
    
    ServerBuilder builder;
    
    // 监听指定地址，不使用认证
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    
    // 注册服务
    builder.RegisterService(&service);
    
    // 构建并启动服务器
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "============================================================" << std::endl;
    std::cout << "UAV gRPC 服务端" << std::endl;
    std::cout << "============================================================" << std::endl;
    std::cout << "监听端口: " << port << std::endl;
    std::cout << "服务器地址: " << server_address << std::endl;
    std::cout << "============================================================" << std::endl;
    std::cout << std::endl;
    std::cout << "[服务端] gRPC 服务器已启动，等待客户端连接..." << std::endl;
    
    // 等待服务器关闭
    server->Wait();
}

// ============================================================================
// 主函数
// ============================================================================

int main(int argc, char** argv) {
    int port = 50051;
    
    // 解析命令行参数
    if (argc > 1) {
        if (std::string(argv[1]) == "--port" && argc > 2) {
            port = std::stoi(argv[2]);
        } else if (std::string(argv[1]) == "--help" || std::string(argv[1]) == "-h") {
            std::cout << "用法: " << argv[0] << " [--port <端口>]" << std::endl;
            std::cout << "默认端口: 50051" << std::endl;
            return 0;
        }
    }
    
    RunServer(port);
    
    return 0;
}

File: example/python/grpc_server.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
gRPC 服务端示例
演示如何使用生成的 protobuf 和 gRPC 代码实现服务端
"""

import sys
import os
import grpc
from concurrent import futures
import logging

# 添加父目录到路径，以便导入生成的 protobuf 代码
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../..', 'generated', 'python'))

try:
    import uav_service_pb2
    import uav_service_pb2_grpc
except ImportError as e:
    print(f"错误: 无法导入生成的 gRPC 代码: {e}")
    print("请先运行 'make python' 生成 gRPC 代码")
    print("如果已生成，请确保已安装 grpcio: pip install grpcio")
    sys.exit(1)

# 配置日志
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class UavServiceServicer(uav_service_pb2_grpc.UavServiceServicer):
    """实现 UavService 服务接口"""
    
    def UploadStatus(self, request, context):
        """
        处理上传无人机状态数据的请求
        
        Args:
            request: UploadStatusRequest 消息
            context: gRPC 上下文
            
        Returns:
            UploadStatusResponse 消息
        """
        logger.info(f"收到上传状态请求 - UAV ID: {request.uav_id}, 数据长度: {len(request.data)} 字节")
        
        # 在这里添加你的业务逻辑
        # 例如：验证数据、保存到数据库、处理状态信息等
        try:
            # 示例：简单的数据验证
            if not request.uav_id:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details("uav_id 不能为空")
                return uav_service_pb2.UploadStatusResponse(
                    success=False,
                    message="uav_id 不能为空"
                )
            
            if not request.data:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details("data 不能为空")
                return uav_service_pb2.UploadStatusResponse(
                    success=False,
                    message="data 不能为空"
                )
            
            # TODO: 在这里添加你的业务逻辑
            # 例如：
            # - 解析和验证数据
            # - 保存到数据库
            # - 更新无人机状态
            # - 触发其他服务
            
            logger.info(f"成功处理 UAV {request.uav_id} 的状态数据")
            
            return uav_service_pb2.UploadStatusResponse(
                success=True,
                message=f"成功接收 UAV {request.uav_id} 的状态数据，数据大小: {len(request.data)} 字节"
            )
            
        except Exception as e:
            logger.error(f"处理上传状态请求时出错: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"服务器内部错误: {str(e)}")
            return uav_service_pb2.UploadStatusResponse(
                success=False,
                message=f"服务器错误: {str(e)}"
            )
    
    def SetSafetySpace(self, request, context):
        """
        处理下发无人机安全空间的请求
        
        Args:
            request: SetSafetySpaceRequest 消息
            context: gRPC 上下文
            
        Returns:
            SetSafetySpaceResponse 消息
        """
        logger.info(f"收到设置安全空间请求 - UAV ID: {request.uav_id}, 数据长度: {len(request.data)} 字节")
        
        # 在这里添加你的业务逻辑
        # 例如：验证安全空间数据、下发到无人机、更新配置等
        try:
            # 示例：简单的数据验证
            if not request.uav_id:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details("uav_id 不能为空")
                return uav_service_pb2.SetSafetySpaceResponse(
                    success=False,
                    message="uav_id 不能为空"
                )
            
            if not request.data:
                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
                context.set_details("data 不能为空")
                return uav_service_pb2.SetSafetySpaceResponse(
                    success=False,
                    message="data 不能为空"
                )
            
            # TODO: 在这里添加你的业务逻辑
            # 例如：
            # - 解析和验证安全空间数据
            # - 检查权限
            # - 下发到无人机
            # - 更新配置数据库
            # - 通知相关服务
            
            logger.info(f"成功为 UAV {request.uav_id} 设置安全空间")
            
            return uav_service_pb2.SetSafetySpaceResponse(
                success=True,
                message=f"成功为 UAV {request.uav_id} 设置安全空间，数据大小: {len(request.data)} 字节"
            )
            
        except Exception as e:
            logger.error(f"处理设置安全空间请求时出错: {e}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"服务器内部错误: {str(e)}")
            return uav_service_pb2.SetSafetySpaceResponse(
                success=False,
                message=f"服务器错误: {str(e)}"
            )


def serve(port=50051):
    """
    启动 gRPC 服务器
    
    Args:
        port: 服务器监听端口，默认 50051
    """
    # 创建 gRPC 服务器
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    
    # 注册服务
    uav_service_pb2_grpc.add_UavServiceServicer_to_server(
        UavServiceServicer(), server
    )
    
    # 监听端口
    listen_addr = f'[::]:{port}'
    server.add_insecure_port(listen_addr)
    
    # 启动服务器
    server.start()
    logger.info(f"gRPC 服务器已启动，监听端口: {port}")
    logger.info(f"服务器地址: {listen_addr}")
    
    try:
        # 保持服务器运行
        server.wait_for_termination()
    except KeyboardInterrupt:
        logger.info("收到停止信号，正在关闭服务器...")
        server.stop(0)
        logger.info("服务器已关闭")


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description='UAV gRPC 服务端')
    parser.add_argument(
        '--port',
        type=int,
        default=50051,
        help='服务器监听端口 (默认: 50051)'
    )
    
    args = parser.parse_args()
    
    print("=" * 60)
    print("UAV gRPC 服务端")
    print("=" * 60)
    print(f"监听端口: {args.port}")
    print("=" * 60)
    print()
    
    serve(port=args.port)

